### PR TITLE
Support world reset

### PR DESCRIPTION
**Pull request by azeey**
# 🎉 New feature

Toward #1107 

## Summary
This adds a new phase in which systems can implement special Reset functionality. The approach taken is to make an initial copy of the ECM after the entities have been created. When reset is requested, a diff is computed between the current ECM and the initial copy. This is necessary to ensure that entities that were added later on are marked as removed after reset. Likewise, entities that were removed later on are marked as newly added after reset.

The physics system has been updated to make use of the Reset phase. The changes were nontrivial because we keep a lot of state in the physics system to help with performance. Hopefully, updating other systems will be more straightforward.

The Reset API allows systems to make changes to the ECM. This can be used to, for example, change the poses of objects in the scene at every reset without having to reload the the world. As an example, I have created a small and very trivial joint position randomizer.

There things that still need to be cleaned up and tests that need to be added, but I wanted open this as a draft PR to get feedback on the approach.
 
## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: include/ignition/gazebo/components/Component.hh [hunk 1]
```diff
@@ -282,7 +282,7 @@ namespace components
 
     /// \brief Clone the component.
     /// \return A pointer to the component.
-    public: virtual std::unique_ptr<BaseComponent> Clone() = 0;
+    public: virtual std::unique_ptr<BaseComponent> Clone() const = 0;
```
Comment (chapulina) [inline include/ignition/gazebo/components/Component.hh:285]: Good call. I think this could be broken into a separate PR.

Comment (mjcarroll) [reply include/ignition/gazebo/components/Component.hh:285]: Included in #1300

**Comment by mjcarroll** [comment]
This will be cleaner when we get #1340 landed and forward ported

**Comment by mjcarroll** [comment]
@osrf-jenkins retest this please

File: src/EntityComponentManagerDiff.hh [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+#include <vector>
+
+namespace ignition
+{
+  namespace gazebo
+  {
+    // Inline bracket to help doxygen filtering.
+    inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE {
+
+    class EntityComponentManagerDiff
```
Comment (chapulina) [inline src/EntityComponentManagerDiff.hh:34]: This class should be documented

**Review by chapulina — COMMENTED** [review]
I believe this PR broke compilation on Windows:

```
[1830.187s] EntityComponentManager_TEST.obj : error LNK2001: unresolved external symbol "public: void __cdecl ignition::gazebo::v7::EntityComponentManagerDiff::ClearRemovedEntities(void)" (?ClearRemovedEntities@EntityComponentManagerDiff@v7@gazebo@ignition@@QEAAXXZ) [C:\Jenkins\workspace\ign_gazebo-pr-win\ws\build\ignition-gazebo7\src\UNIT_EntityComponentManager_TEST.vcxproj]
[1830.187s] EntityComponentManager_TEST.obj : error LNK2001: unresolved external symbol "public: class std::vector<unsigned __int64,class std::allocator<unsigned __int64> > const & __cdecl ignition::gazebo::v7::EntityComponentManagerDiff::AddedEntities(void)const " (?AddedEntities@EntityComponentManagerDiff@v7@gazebo@ignition@@QEBAAEBV?$vector@_KV?$allocator@_K@std@@@std@@XZ) [C:\Jenkins\workspace\ign_gazebo-pr-win\ws\build\ignition-gazebo7\src\UNIT_EntityComponentManager_TEST.vcxproj]
[1830.187s] EntityComponentManager_TEST.obj : error LNK2001: unresolved external symbol "public: class std::vector<unsigned __int64,class std::allocator<unsigned __int64> > const & __cdecl ignition::gazebo::v7::EntityComponentManagerDiff::RemovedEntities(void)const " (?RemovedEntities@EntityComponentManagerDiff@v7@gazebo@ignition@@QEBAAEBV?$vector@_KV?$allocator@_K@std@@@std@@XZ) [C:\Jenkins\workspace\ign_gazebo-pr-win\ws\build\ignition-gazebo7\src\UNIT_EntityComponentManager_TEST.vcxproj]
[1830.187s] C:\Jenkins\workspace\ign_gazebo-pr-win\ws\build\ignition-gazebo7\bin\Release\UNIT_EntityComponentManager_TEST.exe : fatal error LNK1120: 3 unresolved externals [C:\Jenkins\workspace\ign_gazebo-pr-win\ws\build\ignition-gazebo7\src\UNIT_EntityComponentManager_TEST.vcxproj]
```

@azeey / @mjcarroll , can you take a look?
